Create Blog Post “getting-started-with-flutter-3-and-vonage-apis” (#2114)
Nexmo/deved-platform: content/blog/en/getting-started-with-flutter-3-and-vonage-apis.md

Run the following commands, remember to replace the `APPLICATION_ID` variable with the ID of your application and `PRIVATE_KEY` with the name of your private key file.

You are generating a JWT using the Vonage CLI by running the following command but remember to replace the `APP_ID` variable with your own value:

```
vonage jwt --app_id=APPLICATION_ID --subject=Alice --key_file=./PRIVATE_KEY --acl='{"paths":{"/*/users/**":{},"/*/conversations/**":{},"/*/sessions/**":{},"/*/devices/**":{},"/*/image/**":{},"/*/media/**":{},"/*/applications/**":{},"/*/push/**":{},"/*/knocking/**":{},"/*/legs/**":{}}}'
```

The above commands set the expiry of the JWT to one day from now, which is the maximum.

![terminal screenshot of a generated sample JWT](/content/blog/getting-started-with-flutter-3-and-vonage-apis/generated-jwt-key-vonage.png)

We now have everything we need to use the Vonage Voice API within a flutter application. Let's now get the application itself set up.

## Flutter setup

> If you haven't already, start by downloading and installing Flutter and its dependencies. You can do this by following the [Install Guide](https://docs.flutter.dev/get-started/install). Once you have Flutter setup correctly the next thing you will need to do is configure your IDE, how to do this will depend on the IDE you wish to use but the [Set up an editor](https://docs.flutter.dev/get-started/editor) guide will help you with this. 

Reviewer: is there a way to verify flutter setup? e.g. flutter --version. also which version of flutter

> For this guide, we will be using Android Studio. 

Reviewer: 😭

Once your IDE is set up follow the [test drive](https://docs.flutter.dev/get-started/test-drive) guide to set up a basic Flutter application with support for both Android and iOS. We will be using this base app as the start of this project, but of course, if you already have a Flutter project you want to use you can do this as well.

## Installing SDKs

> With the project now set up we can install the Vonage client SDK. Currently, the Client SDK is not available as a Flutter package, so we will have to use the [Android native Client SDK](https://developer.nexmo.com/client-sdk/setup/add-sdk-to-your-app/android) and the [iOS native Client SDK](https://developer.vonage.com/client-sdk/setup/add-sdk-to-your-app/ios) Communicate between Android/iOS and Flutter will use [MethodChannel](https://api.flutter.dev/flutter/services/MethodChannel-class.html) - this way, Flutter will call Android/iOS methods, Android/iOS will call Flutter methods.

Reviewer: ```suggestion
With the project now set up, we can install the Vonage client SDK. Currently, the Client SDK is not available as a Flutter package, so we will use the [Android native Client SDK](https://developer.nexmo.com/client-sdk/setup/add-sdk-to-your-app/android) and the [iOS native Client SDK](https://developer.vonage.com/client-sdk/setup/add-sdk-to-your-app/ios). Communication between Android/iOS and Flutter will use [MethodChannel](https://api.flutter.dev/flutter/services/MethodChannel-class.html) - this way, Flutter will call Android/iOS methods, and Android/iOS will call Flutter methods.
```

```
dependencies {
    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
    implementation "com.nexmo.android:client-sdk:4.1.0"
}
```

Finally, make sure your `minSdkVersion` is set to at least `23`:

```
    defaultConfig {
        applicationId "com.vonage.tutorial.voice.app_to_phone"
        minSdkVersion 23
        targetSdkVersion flutter.targetSdkVersion
        versionCode flutterVersionCode.toInteger()
        versionName flutterVersionName
    }
```

The Android SDK is now set up and ready to be used for the Android build of the flutter application.

### iOS SDK

To install the iOS SDK start with generating the `PodFile` by opening a command line in the root of your Flutter project and then running the commands below:

```
cd ios/
pod init
```

This will generate the `PodFile`, open this file and add the below pod:

```
pod 'NexmoClient'
```

Make sure to also set the platform to at least ios 10

```
platform :ios, '10.0'
```

Your complete file should look something like this:

```
platform :ios, '11.0'

target 'Runner' do
  use_frameworks!

  pod 'NexmoClient'
end
```

Next from the command line, again in the iOS directory run:

```

> pod update

Reviewer: ```suggestion
pod install
```

```

This will download and install the Vonage SDK and its dependencies.

Finally to link this to your Flutter project, from the root directory of your project run the below Flutter command. This will trigger an iOS build and generate the files needed to make use of the SDK.

```
flutter build ios
```

Once complete and successfully built your SDK is set up and ready to be used.

## Code

> By the nature of Flutter, the code can easily be broken down into three areas, the Flutter code which is written in Dart, the native Android code which is written in Kotlin and the native iOS code which is written in Swift.

Reviewer: ```suggestion
By the nature of Flutter, the code can be broken down into three areas, the Flutter code which is written in Dart, the native Android code which is written in Kotlin and the native iOS code which is written in Swift.
```

          children: <Widget>[
            const SizedBox(height: 64),
            _updateView()
          ],
        ),
      ),
    );
  }

  Widget _updateView() {
    if (_sdkState == SdkState.LOGGED_OUT) {
      return ElevatedButton(
          onPressed: () { _loginUser(); },
          child: const Text("LOGIN AS ALICE")
      );
    } else if (_sdkState == SdkState.WAIT) {
      return const Center(
        child: CircularProgressIndicator(),
      );
    } else if (_sdkState == SdkState.LOGGED_IN) {
      return ElevatedButton(
          onPressed: () { _makeCall(); },
          child: const Text("MAKE PHONE CALL")
      );
    } else if (_sdkState == SdkState.ON_CALL) {
      return ElevatedButton(
          onPressed: () { _endCall(); },
          child: const Text("END CALL")
      );
    } else {
      return const Center(
          child: Text("ERROR")
      );
    }
  }
}

enum SdkState {
  LOGGED_OUT,
  LOGGED_IN,
  WAIT,
  ON_CALL,
  ERROR
}
```

This is the complete class needed to build the app's UI and trigger the platform-specific methods which we will write in a moment. Let's break down what's going on in each of the methods in this class.

> Starting with the imports at the top of this class, we have the normal flutter imports but we are also using the [permission handler](https://pub.dev/packages/permission_handler/install) package. This is used to manage requesting permissions on iOS and Android for us. Make sure you have installed this by running the command:

Reviewer: ```suggestion
Starting with the imports at the top of this class, we have the normal Flutter imports but we are also using the [permission handler](https://pub.dev/packages/permission_handler/install) package. This is used to manage requesting permissions on iOS and Android for us. Make sure you have installed this by running the command:
```

```
flutter pub add permission_handler
```

> At the root of your flutter project.

Reviewer: I would move this before the colon above

Next, we build the app, for this demo we have a very simple app with just one widget element which we have called `CallWidget`

```dart
void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      title: 'Flutter Demo',
      home: CallWidget(title: 'app-to-phone-flutter'),
    );
  }
}
```

> This `CallWidget` extends the `StatefulWidget` taking the title and initialising the `CallWidgetState`.

Reviewer: ```suggestion
This `CallWidget` extends the `StatefulWidget`, taking the title and initialising the `CallWidgetState`.
```

          children: <Widget>[
            const SizedBox(height: 64),
            _updateView()
          ],
        ),
      ),
    );
  }
```

Next, the `_updateView` method is used to change what is currently displayed in the box based on the current state of the app. This state model allows for a clean UI only showing the user what they need to see at any given time in the app life cycle.

```dart
  Widget _updateView() {
    if (_sdkState == SdkState.LOGGED_OUT) {
      return ElevatedButton(
          onPressed: () { _loginUser(); },
          child: const Text("LOGIN AS ALICE")
      );
    } else if (_sdkState == SdkState.WAIT) {
      return const Center(
        child: CircularProgressIndicator(),
      );
    } else if (_sdkState == SdkState.LOGGED_IN) {
      return ElevatedButton(
          onPressed: () { _makeCall(); },
          child: const Text("MAKE PHONE CALL")
      );
    } else if (_sdkState == SdkState.ON_CALL) {
      return ElevatedButton(
          onPressed: () { _endCall(); },
          child: const Text("END CALL")
      );
    } else {
      return const Center(
          child: Text("ERROR")
      );
    }
  }
```

> The _loginUser and _endCall methods are very similar in that all we are doing here is invoking the loginUser/endCall methods in the native code. This is how we trigger the native code when the user presses a button on the UI. Within the `_loginUser` we have a variable `token` this should be the JWT value you generated earlier using the Vonage CLI

Reviewer: ```suggestion
The `_loginUser` and `_endCall` methods are very similar in that all we are doing here is invoking the `loginUser`/`endCall` methods in the native code. This is how we trigger the native code when the user presses a button on the UI. Within the `_loginUser` we have a variable `token` this should be the JWT value you generated earlier using the Vonage CLI
```

```dart
Future<void> _loginUser() async {
    String token = "ALICE_TOKEN";

    try {
      await platformMethodChannel
          .invokeMethod('loginUser', <String, dynamic>{'token': token});
    } on PlatformException catch (e) {
      if (kDebugMode) {
        print(e);
      }
    }
  }

  Future<void> _endCall() async {
    try {
      await platformMethodChannel.invokeMethod('endCall');
    } on PlatformException catch (e) {
      if (kDebugMode) {
        print(e);
      }
    }
  }
```

> The `_makeCall` method also involved a method on the native code, calling the makeCall method. However, before it does that we use the requestPermissions method to request the required run time permissions from the user. In this case that is just the microphone/audio recording.

Reviewer: invoked?

> The `_makeCall` method also involved a method on the native code, calling the makeCall method. However, before it does that we use the requestPermissions method to request the required run time permissions from the user. In this case that is just the microphone/audio recording.

Reviewer: ```suggestion
The `_makeCall` method also involved a method on the native code, calling the `makeCall` method. However, before it does that we use the `requestPermissions` method to request the required run time permissions from the user. In this case that is just the microphone/audio recording.
```

                when (call.method) {
                    "loginUser" -> {
                        val token = requireNotNull(call.argument<String>("token"))
                        loginUser(token)
                        result.success("")
                    }
                    "makeCall" -> {
                        makeCall()
                        result.success("")
                    }
                    "endCall" -> {
                        endCall()
                        result.success("")
                    }
                    else -> {
                        result.notImplemented()
                    }
                }
            }
        }
    }
```

The `loginUser` method is called when Flutter sends the loginUser method call, this passes in the JWT token we set and then triggers the login method on the client.

```kotlin
private fun loginUser(token: String) {
        client.login(token)
    }
```

The `makeCall` method is called when Flutter sends the makeCall method call, this starts a phone call to the specified phone number `"PHONE_NUMBER"` you should replace this with an actual phone number that you wish to call.

> Again here we pass back the state to Flutter depending on if the call is successful and starts or if there is some kind of error.

Reviewer: ```suggestion
Again, here we pass back the state to Flutter depending on if the call is successful and starts or if there is some kind of error.
```

```kotlin
    private fun makeCall() {
        notifyFlutter(SdkState.WAIT)

        client.serverCall("PHONE_NUMBER", null, object : NexmoRequestListener<NexmoCall> {
            override fun onSuccess(call: NexmoCall?) {
                onGoingCall = call
                onGoingCall?.addCallEventListener(callEventListener)
                notifyFlutter(SdkState.ON_CALL)
            }

            override fun onError(apiError: NexmoApiError) {
                notifyFlutter(SdkState.ERROR)
            }
        })
    }
```

> The `endCall` method is called when Flutter sends the endCall method call, this ends the current phone call (if there is one). 

Reviewer: ```suggestion
The `endCall` method is called when Flutter sends the `endCall` method call, this ends the current phone call (if there is one). 
```

```kotlin
    private fun endCall() {
        notifyFlutter(SdkState.WAIT)

        onGoingCall?.hangup(object : NexmoRequestListener<NexmoCall> {
            override fun onSuccess(call: NexmoCall?) {
                onGoingCall?.removeCallEventListener(callEventListener)
                onGoingCall = null
                notifyFlutter(SdkState.LOGGED_IN)
            }

            override fun onError(apiError: NexmoApiError) {
                notifyFlutter(SdkState.ERROR)
            }
        })
    }
```

> Finally, we have the `nofityFlutter` method, this is where we use the Flutter magic to send back the current state of the application so Flutter can update the UI. Using this we are able to involve the Flutter updateState method and pass the current state as a variable.

Reviewer: ```suggestion
Finally, we have the `nofityFlutter` method, this is where we use the Flutter magic to send back the current state of the application so Flutter can update the UI. Using this we are able to involve the Flutter `updateState` method and pass the current state as a variable.
```

Start the Android emulator so that flutter can attach to it, below is where you can do this in Android studio

![Android studio with device manager selected](/content/blog/getting-started-with-flutter-3-and-vonage-apis/screenshot-2022-05-18-at-16.24.37.png)

Once this is running you can select this device as the target for the Flutter build and press the green arrow to build and run the Flutter app (with Android native code).

![android studio with emulator selected and main.dart](/content/blog/getting-started-with-flutter-3-and-vonage-apis/screenshot-2022-05-18-at-16.24.53.png)

Once the application has build and installed you will be presented with the below left screen. Clicking the Login as Alice button will take you to the next screen. From here you can press the Make phone call button which will (on the first run) prompt you to allow the audio permissions. After this the phone call will start and the phone number you entered will be called connecting the audio session.

Once you wish to finish the call you can do so by pressing the end call button.

> ![](/content/blog/getting-started-with-flutter-3-and-vonage-apis/app-screens.png)

Reviewer: missing alt text
